include/ur_api.h

```diff
@@ -2739,6 +2765,41 @@ typedef enum ur_usm_alloc_info_t
 
 } ur_usm_alloc_info_t;
 
+///////////////////////////////////////////////////////////////////////////////
+/// @brief USM allocation descriptor type
+typedef struct ur_usm_desc_t
+{
+    ur_structure_type_t stype;                      ///< [in] type of this structure
```

Comment:
This pattern comes from Vulkan and LevelZero, I think this [SO exmplanation](https://stackoverflow.com/questions/36347236/vulkan-what-is-the-point-of-stype-in-vkcreateinfo-structs/36347301#36347301) for the purpose of `stype` in the presence of `pNext` to create a linked list of extensible structs is a good summary.

Reply:
To work properly though, we need a enum of all the struct which have this extension mechanism, like [VkStructureType](https://github.com/KhronosGroup/Vulkan-Headers/blob/main/include/vulkan/vulkan_core.h#L196). I think this might be worth spinning out into a separate issue since we'll need to properly define what the UR extension mechanisms are in the specification and them apply it consistently.

Reply:
Thanks, that makes sense. So a 'structure type' is some algebraic data type composed of different description types so we can form a linked-list.

Are all the enum values in `ur_structure_type_t` valid for every struct that accepts 'pNext'? Right now this is sharing a type enum with some other functionality. Would it be useful to define a unique base type for each linked-list of parameters? This way we retain type information and it's clear which types go with what list. Like this:
```
enum usm_desc_types {
  USM_DESC_FOO,
  USM_DESC_BAR,
};

struct usm_desc_base {
   enum usm_desc_types type;
   struct usm_desc_base *next;
};

struct usm_desc {
  struct usm_desc_base base;
  ...;
};

struct usm_desc *foo = ...;
struct usm_desc *bar = ...;

foo->base.next = &bar->base;
```

Reply:
> Are all the enum values in ur_structure_type_t valid for every struct that accepts 'pNext'?

I beleive it should be a 1:1 mapping between a single `ur_structure_type_t::<enum>` and a single `ur_<struct>_t`.

Scoping the structure types is interesting 🤔 `VkStructureType` has become very large over time.

Reply:
Yes, I like this idea as well. The only drawback is that it would not be consistent with vulcan/level_zero. But I'm not sure if this is a big problem.

Reply:
Each enum would have all struct types that are the component parts of the list, regardless of whether they are present on other lists or not. That would make it a bit harder to represent this as an enum/disjoint union/datatype in hypothetical bindings for languages with richer type systems, but I'm not sure we should be concerned about that too much.

The benefit is that the type system helps the user and its harder to make a mistake.

Reply:
Ok, that makes sense. I can do that change in a separate PR.

Reply:
I'm still on the fence about this direction. I do like the idea of leveraging the type system but I'm also warey of introducing another extension mechanism, different that others in the same software stack, which we will need to explain to people.

I think we should continue discussion in an issue with a view to arriving at a consensus.